### PR TITLE
Make mean teacher algorithm consider distributed training

### LINKED ISSUE #2705: Multi-GPU training integration test failure
The failure occurs in daily tests for at least a week in a row :
https://github.com/openvinotoolkit/training_extensions/actions/runs/7105336612
FAILED tests/e2e/cli/detection/test_detection.py::TestToolsOTXSemiSLDetection::test_otx_multi_gpu_train_semisl[Object_Detection_ResNeXt101_ATSS]

@eunwoosh could you have a look?

### PR DESCRIPTION
**Pull request by eunwoosh**
### Summary
This PR fixes #2705 
The reason of bug is that when distributed training, one process may train a model w/ pseudo label but other processes doesn't  in some case. In this case, each process executes different code flow, which makes the error.
This PR changes code as below

-  If any process will train a model w/ pseudo label, then all process train a model w/ pseudo label.
- Not to affect backprop, process, which doesn't have to train w/ pseudo label, get 0 loss from pseudo label.

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Comment by eunwoosh** [comment]
Hi @kprokofi. I changed mean teacher algorithm code to make it compatible with distributed training.
It seems that those code is written by you, so could you check this change is ok?
I checked that the error doesn't occur w/ this change but I'm not sure it doesn't affect to training.

File: src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py [hunk 1]
```diff
@@ -182,22 +183,29 @@ def forward_train(
         pseudo_bboxes, pseudo_labels, pseudo_masks, pseudo_ratio = self.generate_pseudo_labels(
             teacher_outputs, device=current_device, img_meta=ul_img_metas, **kwargs
         )
+        non_empty = [bool(len(i)) for i in pseudo_labels] if self.filter_empty_annotations else [True]
+        get_unlabeled_loss = pseudo_ratio >= self.min_pseudo_label_ratio and any(non_empty)
+
+        if dist.is_initialized():
+            reduced_get_unlabeled_loss = torch.tensor(int(get_unlabeled_loss)).to(current_device)
```
Comment (kprokofi) [inline src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py:190]: Why do we need to wrap the bool variable to tensor and set the device?

File: src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py [hunk 2]
```diff
@@ -182,22 +183,29 @@ def forward_train(
...
+            dist.all_reduce(reduced_get_unlabeled_loss)
+            if dont_have_to_train := not get_unlabeled_loss and reduced_get_unlabeled_loss > 0:
```
Comment (kprokofi) [inline src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py:192]: If dont_have_to_train is True -> then we don't need to compute unlabeled loss. Why do we set get_unlabeled_loss to True if it was False before? Can we avoid such behavior, it is too complicated to understand when read the code.

File: src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py [hunk 3]
```diff
@@ -182,22 +183,29 @@ def forward_train(
...
+                get_unlabeled_loss = True
+                non_empty[0] = True
+
         if self.filter_empty_annotations:
```
Comment (kprokofi) [inline src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py]: Please, put this block inside "if get_unalbeled_loss" statement, since we don't need to compute this when pseudo_ratio < self.min_pseudo_label_ratio. I know it wasn't written by you, but it just catches my eye.

Comment (eunwoosh) [reply src/otx/algorithms/detection/adapters/mmdet/models/detectors/mean_teacher.py]: Thanks for comment. But it seems that if this statement is moved into `if get_unlabeled_loss`, which argumentr are given to `self._visual_online()` can be different, when `get_unlabeled_loss` is `True` and `self.visualize` is `True`. Is it ok?

**Comment by kprokofi** [comment]
@eunwoosh Let's merge it as is with this workaround. I think we can return to the refactoring later, when adding this feature to the new OTX2.0 design

**Review by kprokofi — APPROVED** [review]

**Review by sovrasov — APPROVED** [review]
